aten/src/ATen/cuda/PhiloxRNGEngine.h

```diff
@@ -0,0 +1,286 @@
+#pragma once
+
+#include <math.h>
+#include <stdint.h>
+#include <cuda.h>
+#include <c10/macros/Macros.h>
+
+namespace at {
+namespace cuda {
+
+/*
+* Philox Engine implementation
+* Originally implemented in PyTorch's fusion compiler
+* Refer to: https://ieeexplore.ieee.org/document/6114424
+* for details regarding the engine.
+*
+* The Philox engine is currently used in CUDA distributions
+* kernels as its random engine. 
+* 
+* It takes a seed value, a subsequeunce
+* for starting the generation and an offset for the sequence.
+*
+* Think of this engine as an algorithm producing a huge array. We are 
+* parallelizing this array by partitioning the huge array and assigning 
+* a thread index to each partition. In other words, each seed value 
+* (there are 2^64 possible seed values) gives a sub array of size 
+* 2^128 (each element in that array is a 128 bit number). Reasoning
+* behind the array being of size 2^128 is, there are 2^64 possible
+* thread index value and there is an array of size 2^64 for each of
+* those thread index. Hence 2^64 * 2^64 = 2^128 for each seed value.
+*
+* In short, this generator can produce 2^64 (seed values) * 2^128 (number
+* of elements in an array given by a seed value) = 2^192 values.
+*
+* Arguments:
+* seed:        Seed values could be any number from 0 to 2^64-1.
+* subsequence: Subsequence is just the cuda thread indexing with:
+*              - blockIdx.x * blockDim.x + threadIdx.x
+* offset:      The offset variable in PhiloxEngine  decides how many 128-bit 
+*              random numbers to skip (i.e. how many groups of 4, 32-bit numbers to skip)
+*              and hence really decides the total number of randoms that can be achieved 
+*              for the given subsequence. NOTE: there is a hard requirement for this
```

Comment:
> there is a hard requirement for this number to be a multiple of 4
Is that true?  If "offset" defines which set-of-128-bits you are producing (which appears to be the case) then "offset" can be anything, it doesn't need to be a multiple of 4.